### ISSUE TITLE
hyper-v "time synchronization" integration not working in hyper-v image (sleep/resume problem)?

### ISSUE THREAD

**Issue by Andy2244** [clearlinux/distribution#39]
I noticed that even if i enable "time synchronization" in the hyper-v integration host options, after i wake my VM host from a sleep state, the date/time is not updated for the VM and the VM has the date/time from when the host was put into sleep mode. At some point the network time synchronization kicks in, but this often takes to long and leaves the system in a "bad" date state. So i have to manually restart the network-time service, to quickly get the date/time updated.

As i understand it the "time synchronization" integration feature should take care of this even without any network-time service. So it seems the hyper-v image is not correctly using or has support for this feature?

**Comment by iphutch** [comment]
@bwarden could you take a look to see if this is a Clear issue vs a hyper-v bug.

**Comment by bwarden** [comment]
From what I can see, everything _should_ be in place. Please make sure you've installed the os-cloudguest-azure bundle for the userspace utilities, and follow this guide to verify that the kernel modules are installed properly:
https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/manage/manage-hyper-v-integration-services#start-and-stop-an-integration-service-from-a-linux-guest

Please also provide your client kernel version (uname -r) so I can make sure I'm looking at the right kernel build.

**Comment by bwarden** [comment]
lsusb is OK -- we actually build hv_utils into the kernel statically.

Looks like the services are running in the Clear Linux guest. The Get-Service step above applies to Windows guest VMs. Did you follow this section (from the Hyper-V manager) to ensure they're enabled on the host?

https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/manage/manage-hyper-v-integration-services#turn-an-integration-service-on-or-off-using-hyper-v-manager

With the 21010 image, and those enabled on my system, I see time update automatically whenever I resume the VM.

**Comment by bwarden** [comment]
You can also try "sudo journalctl | grep hv_" to see the status of the daemons.

**Comment by Andy2244** [comment]
@bwarden So the "Stopped  vmictimesync" is only used for windows guests? Yet in the documentation they don't actually provide which service/daemon is responsible for time/shutdown on the linux side?

Here is the journalctl and the date still lists the last shutdown time (1:14), while i typed this command at 10:18 CEST. So i only get a time update via NTP once it kicks in at its normal update interval, without it i wont get any valid time at all.

```
Wed Apr 25 01:14:24 CEST 2018
root@clear-vm ~ # sudo journalctl | grep hv_
Apr 24 10:54:30 clear-vm kernel: hv_utils: Shutdown request received - graceful shutdown initiated
Apr 24 10:54:54 clear-vm kernel: calling  netvsc_drv_init+0x0/0x1000 [hv_netvsc] @ 163
Apr 24 10:54:54 clear-vm kernel: hv_vmbus: registering driver hv_netvsc
Apr 24 10:54:54 clear-vm kernel: initcall netvsc_drv_init+0x0/0x1000 [hv_netvsc] returned 0 after 65 usecs
Apr 24 10:54:55 clear-vm hv_vss_daemon[210]: Hyper-V VSS: VSS starting; pid is:210
Apr 24 10:54:55 clear-vm hv_vss_daemon[210]: Hyper-V VSS: open /dev/vmbus/hv_vss failed; error: 2 No such file or dir                                       ectory
Apr 24 10:54:55 clear-vm kernel: hv_utils: KVP IC version 4.0
Apr 24 10:55:39 clear-vm kernel: hv_balloon: Max. dynamic memory size: 2560 MB
Apr 24 10:58:50 clear-vm kernel: hv_utils: Shutdown IC version 3.0
Apr 24 10:58:52 clear-vm kernel: hv_utils: TimeSync IC version 4.0
Apr 24 10:58:50 clear-vm kernel: hv_utils: Heartbeat IC version 3.0
Apr 24 10:58:50 clear-vm kernel: hv_utils: FCopy IC version 1.1
Apr 24 11:05:28 clear-vm kernel: hv_utils: Shutdown request received - graceful shutdown initiated
Apr 24 11:05:54 clear-vm kernel: calling  netvsc_drv_init+0x0/0x1000 [hv_netvsc] @ 151
Apr 24 11:05:54 clear-vm kernel: hv_vmbus: registering driver hv_netvsc
Apr 24 11:05:54 clear-vm kernel: initcall netvsc_drv_init+0x0/0x1000 [hv_netvsc] returned 0 after 382 usecs
Apr 24 11:05:55 clear-vm hv_vss_daemon[214]: Hyper-V VSS: VSS starting; pid is:214
Apr 24 11:05:55 clear-vm hv_vss_daemon[214]: Hyper-V VSS: open /dev/vmbus/hv_vss failed; error: 2 No such file or dir                                       ectory
Apr 24 11:05:55 clear-vm kernel: hv_utils: KVP IC version 4.0
Apr 24 11:06:39 clear-vm kernel: hv_balloon: Max. dynamic memory size: 2560 MB
Apr 24 11:08:25 clear-vm kernel: hv_utils: Shutdown request received - graceful shutdown initiated
Apr 24 11:08:48 clear-vm kernel: calling  netvsc_drv_init+0x0/0x1000 [hv_netvsc] @ 157
Apr 24 11:08:48 clear-vm kernel: hv_vmbus: registering driver hv_netvsc
Apr 24 11:08:48 clear-vm kernel: initcall netvsc_drv_init+0x0/0x1000 [hv_netvsc] returned 0 after 42 usecs
Apr 24 11:08:49 clear-vm kernel: hv_utils: KVP IC version 4.0
Apr 24 11:08:49 clear-vm hv_vss_daemon[214]: Hyper-V VSS: VSS starting; pid is:214
Apr 24 11:08:49 clear-vm hv_vss_daemon[214]: Hyper-V VSS: open /dev/vmbus/hv_vss failed; error: 2 No such file or dir                                       ectory
Apr 24 11:09:33 clear-vm kernel: hv_balloon: Max. dynamic memory size: 2560 MB
```

The "clear-vm kernel: hv_utils: TimeSync IC version 4.0" line would indicate that it has support for it, yet i get no update. "Get-VMIntegrationService" shows that its offered to the VM from the host, so what else can i try?

**Comment by Andy2244** [comment]
I get the same behavior for ubuntu 17.10 on a different host, so it seems this is a general problem utilizing the LIS time service under linux.

I did dig a little deeper and going by this pdf for the latest [LIS](https://www.microsoft.com/en-us/download/details.aspx?id=55106) the hyperv Time-Service needs to-be properly configured to-be used as time source.

Going by the docs the timesource is installed and working via kernel.
```
 ls /sys/class/ptp
ptp0
```

```
cat /sys/class/ptp/ptpx/clock_name 
hyperv
```
The simplest option suggested is to disable ntp and switch to the timesync source via:
`echo Y > /sys/module/hv_utils/parameters/timesync_mode `

Yet this does not work on clear or ubuntu, since the "parameters" is not present under ubuntu (/sys/module/hv_utils/parameters), while in clear "/sys/module/hv_utils" is not present at all?


The more complex option suggested is to use chronyd instead of ntpd, since the later does not support the ptp source.
```
/etc/chrony.conf: 
refclock PHC /dev/ptp0 poll 3 dpoll -2 offset 
```
I tried this under ubuntu and clear and enabled the service and restarted it. Yet under clear i'm still unsure how are simple config changes handled? Do i copy the /use/share/defaults/chrony/chrony.conf to /etc/chrony.conf and add the extra parameter or do i just create the /etc/chrony.conf and somehow its merged with the default?

I tried both and on ubuntu/clear i get:
```
chronyc sources
210 Number of sources = 5
MS Name/IP address         Stratum Poll Reach LastRx Last sample
===============================================================================
#~ PHC0                          0   3     7     9  -145.5s[-145.5s] +/-  349ns
```

Yet in both systems i get no time updates after a sleep/resume cycle. I would prefer the "echo Y > /sys/module/hv_utils/parameters/timesync_mode " option, yet i have no clue how to enable this. At this point i also have no idea why the chrony approach does not work and i would have to test this all under the official supported centOS as well and verify/cross-reference both solution's.

Maybe someone with a better understanding of this can take a look, since this all would suggest that time synchronization is "broken" for all hyperv images regarding sleep/resume.

PS: My expectations for a hyperv image would be that this actually works by default. So in a ideal world it would be enabled by default for the hyperv images or as a work around systemd needs to update via ntp on a resume state.

**Comment by bwarden** [comment]
Let me rephrase. In addition to checking on the host via "Get-VMIntegrationService", could you please try from the Hyper-V Manager GUI, under Settings for the VM, verifying that Time synchronization is checked under Integration Services? I know it _should_ be the same, but something's not right.

I've tried replicating this on my own system, and I left `date` running in a loop, then saved the VM yesterday afternoon. When I resumed it this morning, the time rolled over exactly as expected. Given that you're having trouble with multiple client VMs, it's more likely that something's not quite right on your host.

As I mentioned, hv_utils is built into the kernel, not as a standalone module. This is why you don't see it in /sys/module. Time sync should work out of the box, without having to pass an extra parameter or configure anything in userspace.

I do have one other idea. By default, we include a user-space SNTP client, `systemd-timesyncd`. In my environment, it can't reach any NTP servers, so it does nothing. If it could reach a server, maybe it could interfere with LIS. You can check its status with `timedatectl`, and if it shows Network Time or NTP synchronized as `yes`, we could try some additional actions. Also, if you have manually configured any other time services, please make sure they're disabled. Having multiple time services trying to set the clock (with their own heuristics for slewing vs. stepping) can cause a lot of problems.

**Comment by Andy2244** [comment]
@bwarden Yes the service is set enabled (checked) from the management gui. I also have a working NTP aka: 
```
timedatectl:
Network time on: yes
NTP synchronized: yes
RTC in local TZ: no
```
I just noticed something: 
> then saved the VM yesterday afternoon. When I resumed it this morning, the time rolled over exactly as expected. 
I'm not manually saveing the VM, i let the host windows system go into its normal sleep/suspend/hibernate state, while the VM is running and than just resume the host.

**Comment by bwarden** [comment]
As an interesting data point, I setup chrony to track the PTP clock device exposed by LIS. This is merely a workaround, but it's less clunky than running an NTP client in a guest VM.

I put this in /etc/chrony.conf (which completely overrides the system defaults from /usr/share/defaults/chrony/chrony.conf):
`refclock PHC /dev/ptp0`

I disabled systemd-timesyncd (the SNTP client):
`systemctl mask --now systemd-timesyncd`

I enabled chrony:
`systemctl enable --now chronyd`

With chrony running, I suspended, waited, and resumed my host. While the time initially didn't match, chrony noticed the disparity and gracefully accelerated the clock to make up lost time within a couple of minutes, as verified in the logs, via:
`journalctl -u chronyd`

**Comment by bwarden** [comment]
You could also add this to chrony.conf to make it step immediately on errors larger than one second:
`makestep 1 -1`

I found [anecdotes like this](https://superuser.com/questions/1312335/why-doesnt-my-debian-hyper-v-guest-update-time-after-host-sleeping) from people with similar experiences, but I haven't found any official documentation. I would guess that it's just assumed you would suspend a guest VM properly before suspending or shutting down a host.

**Comment by Andy2244** [comment]
@bwarden Thanks for looking into this issue, i could reproduce your fix and was apparently just missing the "makestep 1 -1" option to allow the large changes. 
I now get:
```
System clock wrong by 31939.769192 seconds, adjustment started
System clock was stepped by 31939.769192 seconds
```

I guess i will close this issue, since its a LIS/Hyperv specific "oddity".
I still think its strange that the shutdown LIS service transparently correctly handles host restart/shutdowns, without any manual interactions and can even spin-up the VM again if configured on a restart, yet seems to wrongly handle sleep/suspend. It seems to me that having LIS and a direct communication channel to the host, should be enough to handle also those scenarios, but what do i know 😄 

Thanks again for the time diagnosing this and the fix.

**Comment by bwarden** [comment]
I left my host hibernated overnight, and with an error of 54409 seconds, chrony didn't believe the reference clock was accurate, so it didn't update. I'd recommend adding "trust" to the refclock statement so that chrony _always_ believes this clock.

**Comment by Andy2244** [comment]
@bwarden Just a quick followup, i had to add ntp servers also, since after 3+ days of having the vm suspended the clock would not forward again. Chrony seemed not to be able to use/find the refclock device for whatever reasons.

Here is my current config that seem to work for anyone having this issues.
`/etc/chrony.conf`
```
refclock PHC /dev/ptp0 trust poll 2
makestep 1 -1
maxdistance 16.0
pool pool.ntp.org iburst
driftfile /var/lib/chrony/drift
```

**Comment by bwarden** [comment]
Is there any useful information in dmesg or the journal? I wonder if there's an issue with the virtual device.

**Comment by Andy2244** [comment]
@bwarden Forgot to check those sorry.

**Comment by DonJianguo** [comment]
Hello. If only taking RTC/PHC as the source in chrony.conf, what is the confluence on Time Cycle in the VM / HyperV ?
